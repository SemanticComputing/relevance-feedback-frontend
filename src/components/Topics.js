import React from 'react';
import { func, object } from 'prop-types';
import { map, reject, sortBy, reverse } from 'lodash';
import TopicPopover from './TopicPopover';

const Topics = ({ result, describeTopic, currentTopic }) => {

  const topicColors = [
    'green',
    'red',
    'yellow',
    'blue',
    'teal',
    'orange',
    'black',
    'purple',
    'fuchsia',
    'brown',
    'darkgrey',
    'olive',
    'burlywood',
  ];

  const getSize = (topic) => Math.min(24, topic * 30);

  // Sort topics by "size" (i.e. relevance)
  const sortedTopics = reverse(sortBy(
    reject(map(result.topic, (topic, index) => ({ topic, index, size: getSize(topic) })),
      (topic) => topic.topic < 0.1),
    'size'));

  const isSelected = (topic) => (
    currentTopic.result === result.url && currentTopic.topic === topic
  );

  let x = '50%';
  let y = 0;
  let prevSize = 0;

  const circles = map(sortedTopics, ({ index, size }) => {
    y += size + prevSize + 4;
    prevSize = size;

    const selected = isSelected(index);
    const color = topicColors[index];

    const cid = `${result.url}-${index}`;

    return (
      <g key={index}>
        <circle id={cid}
          onMouseOver={() => describeTopic({ result: result.url, topic: index })}
          onMouseLeave={() => describeTopic({})}
          cx={x} cy={y} r={size} fill={color} stroke={selected ? 'black' : color}
          strokeWidth="2" />
        <TopicPopover topic={currentTopic} target={cid}
          isOpen={selected} />
      </g>
    );
  });

  return (
    <svg width="50">
      {circles}
    </svg>
  );
};

Topics.propTypes = {
  result: object.isRequired,
  describeTopic: func.isRequired,
  currentTopic: object.isRequired
};

export default Topics;
