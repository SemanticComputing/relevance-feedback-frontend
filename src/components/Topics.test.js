import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import * as reactstrap from 'reactstrap';
import Topics from './Topics';

const describeTopic = jest.fn();

const result = {
  name: 'title',
  url: 'http://example.fi',
  description: 'description',
  topic: [0.2, 0.1489, 0.1, 0.01]
};

describe('Topics', () => {

  it('renders without crashing', () => {
    const div = document.createElement('div');
    document.body.appendChild(div);
    ReactDOM.render(<Topics describeTopic={describeTopic} result={result} currentTopic={{}} />, div);
    document.body.removeChild(div);
  });

  it('renders topics', () => {
    // Mock the Popover as it requires the DOM and is not relevant here.
    reactstrap.Popover = jest.fn(() => null); // eslint-disable-line import/namespace

    const tree = renderer
      .create(<Topics describeTopic={describeTopic} result={result} currentTopic={{}} />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
