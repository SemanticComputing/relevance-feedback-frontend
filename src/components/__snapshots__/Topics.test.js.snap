// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Topics renders topics 1`] = `
<svg
  width="50"
>
  <g>
    <circle
      cx="50%"
      cy={10}
      fill="green"
      id="http://example.fi-0"
      onMouseLeave={[Function]}
      onMouseOver={[Function]}
      r={6}
      stroke="green"
      strokeWidth="2"
    />
  </g>
  <g>
    <circle
      cx="50%"
      cy={24.467}
      fill="red"
      id="http://example.fi-1"
      onMouseLeave={[Function]}
      onMouseOver={[Function]}
      r={4.4670000000000005}
      stroke="red"
      strokeWidth="2"
    />
  </g>
  <g>
    <circle
      cx="50%"
      cy={35.934}
      fill="yellow"
      id="http://example.fi-2"
      onMouseLeave={[Function]}
      onMouseOver={[Function]}
      r={3}
      stroke="yellow"
      strokeWidth="2"
    />
  </g>
</svg>
`;
